src/controllers/DynamicCategoryFacetQueryController.ts

```diff
@@ -0,0 +1,86 @@
+import { CategoryFacet } from '../ui/CategoryFacet/CategoryFacet';
+import { Assert } from '../misc/Assert';
+import { QueryBuilder } from '../ui/Base/QueryBuilder';
+import { IFacetRequest, IFacetRequestValue } from '../rest/Facet/FacetRequest';
+import { QueryEvents } from '../events/QueryEvents';
+import { FacetValueState } from '../rest/Facet/FacetValueState';
+
+// TODO: Create a real class and move in a separate file
+export type CategoryFacetValue = {
+  value: string;
+  displayValue: string;
+  state: FacetValueState;
+  numberOfResults: number;
+  moreValuesAvailable: boolean;
+  preventAutoSelect: boolean;
+  children: CategoryFacetValue[];
+}
+
+// TODO: rename to simply CategoryFacetQueryController
+export class DynamicCategoryFacetQueryController {
+  private numberOfValuesToRequest: number;
+  private freezeFacetOrder = false;
+
+  constructor(private facet: CategoryFacet) {
+    this.resetNumberOfValuesToRequest();
+    this.resetFreezeCurrentValuesDuringQuery();
+  }
+
+  private resetFreezeCurrentValuesDuringQuery() {
+    this.facet.bind.onRootElement(QueryEvents.duringQuery, () => {
+      this.freezeFacetOrder = false;
+    });
+  }
+
+  public increaseNumberOfValuesToRequest(additionalNumberOfValues: number) {
+    this.numberOfValuesToRequest += additionalNumberOfValues;
+  }
+
+  public resetNumberOfValuesToRequest() {
+    this.numberOfValuesToRequest = this.facet.options.numberOfValues;
+  }
+
+  public putFacetIntoQueryBuilder(queryBuilder: QueryBuilder) {
+    Assert.exists(queryBuilder);
+
+    queryBuilder.facetRequests.push(this.facetRequest);
+    if (this.freezeFacetOrder) {
+      queryBuilder.facetOptions.freezeFacetOrder = true;
```

Comment:
Would be possible to set the queryBuilder freezeFacetOrder value using the property of the same name?

```
queryBuilder.facetOptions.freezeFacetOrder = this.freezeFacetOrder
```